src/impy/models/pythia8.py

```diff
@@ -1,256 +1,157 @@
+            f"Random:seed = {self._seed}",
             # Specify energy in center of mass
             "Beams:frameType = 1",
-            # Minimum bias events
-            "SoftQCD:all = on",
+            "SoftQCD:inelastic = on",
         ]
-
         # Add more options from config file
-        for param_string in impy_config["pythia8"]["options"]:
-            info(5, "Using Pythia 8 parameter:", param_string)
-            self.save_init_strings.append(param_string)
+        config += impy_config["pythia8"]["options"]
 
-        self.attach_log(fname=logfname)
-
-        # Create history for particle decay settings
-        # since changing energy changes resets the stable settings
-        self.stable_history = {}
+        if k.p1_is_nucleus or k.p2_is_nucleus:
+            import warnings
 
-        # self._set_event_kinematics(event_kinematics)
+            warnings.warn(
+                "Support for nuclei is experimental; event generation takes a long time",
+                RuntimeWarning,
+            )
 
-    def set_stable(self, pdgid, stable=True):
+            # speed-up initialization by not fitting nuclear cross-section
+            config.append("HeavyIon:SigFitNGen = 0")
+            config.append("HeavyIon:SigFitDefPar = 10.79,1.75,0.30,0.0,0.0,0.0,0.0,0.0")
 
-        may_decay = not stable
-        if self.lib is not None:
-            self.lib.particleData.mayDecay(pdgid, may_decay)
+        config += [
+            f"Beams:idA = {k.p1pdg}",
+            f"Beams:idB = {k.p2pdg}",
+            f"Beams:eCM = {k.ecm}",
+        ]
 
-        info(5, pdgid, "allowed to decay: ", may_decay)
+        for line in config:
+            if not pythia.readString(line):
+                raise RuntimeError(f"readString('{line}') failed")
 
-        self.stable_history[pdgid] = stable
+        # calling init several times is allowed
+        if not pythia.init():
+            raise RuntimeError("Pythia8 initialization failed")
 
-    def generate_event(self):
-        return not self.lib.next()
+    def _attach_log(self, fname):
+        # not implemented
+        pass
 
+    def _set_stable(self, pdgid, stable):
+        self._lib.pythia.particleData.mayDecay(pdgid, not stable)
 
-class Pythia8(PYTHIA8Run):
-    def __init__(self, event_kinematics, seed="random", logfname=None):
-        from impy.definitions import interaction_model_by_tag as models_dict
+    def _generate_event(self):
+        import warnings
 
-        interaction_model_def = models_dict["PYTHIA8"]
-        super().__init__(interaction_model_def)
-        self.init_generator(event_kinematics, seed, logfname)
+        success = self._lib.pythia.next()
+        if not success:
+            warnings.warn(
+                "PYTHIA8 event rejected. If this warning appears too frequently, "
+                + "the event generator might be not properly configured or "
+                + "used outside of the valid range",
+                RuntimeWarning,
+            )
```

Comment:
This message should go into MCRun, which already contains logic to retry generation and logs the number of retries to detect misconfiguration for all models.

Reply:
I'm not confident putting into `MCEvent` because all other models don't really have problems. And if a model rejects each 100th event, which is not of a big concern it should not spam the STDOUT with the warning all the time. I could add a counter but who knows if a user wants to generate 1000000 events. Then still one would have 10000 warnings in the STDOUT.

Reply:
The cool thing about Warnings is that they are generated exactly once, even if warn is called multiple times. So it does not spam the output.

Reply:
If you look into MCRun.__call__, we already do some logging here if a model needs many retries. What I am suggesting is to put the warning here instead of or in addition to using `info(...)`.

Reply:
OK, great. I put a warning if after 10 rejections have occurred.

Reply:
> Actually, not so simple to add this into MCEvent. We don't want to sum rejections on a class variable for all models.

I cannot follow what kind of problem you have in mind. There are class variables and instances variables. class variables are shared between all instances, and instance variables are specific to each instance. If an instance variable is defined in a base class, it is still local to the instance which is created from a derived class.

The current code `MCRun.__call__` uses a local variable `ntrials` to detect that a model had to retry several times. I think that's good enough. If you call the generator with nevents=1 many times, you will never see it, but that's not the usual use case.

You can also an instance variable, e.g. self._ntrials to count the trials correctly over multiple calls to `MCRun.__call__`.

Reply:
Documented as issue #85 for future improvement.